File: FinalProject/app/src/main/java/com/example/finalproject/fragments/InventoryFragment.java
package com.example.finalproject.fragments;

import android.graphics.Bitmap;
import android.graphics.Point;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.util.Pair;
import android.view.Display;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.example.finalproject.items.Item;
import com.example.finalproject.R;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

public class InventoryFragment extends Fragment {
    private TextView noItems;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_inventory, container, false);
    }

    @Override
    public void onViewCreated(@NonNull @NotNull View view, @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Display display = getActivity().getWindowManager().getDefaultDisplay();
        Point size = new Point();
        display.getSize(size);
        int width = size.x;
        noItems = getView().findViewById(R.id.textView12);
        RecyclerView inventory = (RecyclerView) getView().findViewById(R.id.list);
        ArrayList<ImageView> categories = new ArrayList<>();
        Button back = (Button) getView().findViewById(R.id.inventory_back_button);
        View.OnClickListener clickListener = new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                inventory.setAdapter(new InventoryAdapter(MainActivity.player.getInventory(), categories.indexOf(v)));
            }
        };
        Bitmap bm = Bitmap.createScaledBitmap
                (MainActivity.textures[6][0], MainActivity.categoryImageWidth, MainActivity.categoryImageWidth, false);
        categories.add(getView().findViewById(R.id.armor_weapons));
        categories.add(getView().findViewById(R.id.potions_food));
        categories.add(getView().findViewById(R.id.recourses));
        categories.add(getView().findViewById(R.id.other));
        categories.get(0).setImageBitmap(Bitmap.createBitmap(bm));
        bm = Bitmap.createScaledBitmap
                (MainActivity.textures[6][1], MainActivity.categoryImageWidth, MainActivity.categoryImageWidth, false);
        categories.get(1).setImageBitmap(Bitmap.createBitmap(bm));
        bm = Bitmap.createScaledBitmap
                (MainActivity.textures[6][2], MainActivity.categoryImageWidth, MainActivity.categoryImageWidth, false);
        categories.get(2).setImageBitmap(Bitmap.createBitmap(bm));
        bm = Bitmap.createScaledBitmap
                (MainActivity.textures[6][3], MainActivity.categoryImageWidth, MainActivity.categoryImageWidth, false);
        categories.get(3).setImageBitmap(Bitmap.createBitmap(bm));
        inventory.setAdapter(new InventoryAdapter(MainActivity.player.getInventory()));
        inventory.setLayoutManager(new LinearLayoutManager(getContext()));
        categories.get(0).setOnClickListener(clickListener);
        categories.get(1).setOnClickListener(clickListener);
        categories.get(2).setOnClickListener(clickListener);
        categories.get(3).setOnClickListener(clickListener);
        back.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                FragmentManager fm = getParentFragmentManager();
                FragmentTransaction fragmentTransaction = fm.beginTransaction();
                fragmentTransaction.remove(fm.findFragmentById(R.id.inventory));
                fragmentTransaction.add(R.id.map, new MapFragment(MainActivity.player.getMapNum()));
                fragmentTransaction.add(R.id.status, new StatusBarFragment());
                fragmentTransaction.add(R.id.menu, new MenuFragment());
                fragmentTransaction.commit();
            }
        });
    }

    private class InventoryAdapter extends RecyclerView.Adapter<InventoryAdapter.ViewHolder> {
        private ArrayList<Pair<Item, Integer>> data = new ArrayList<>();
        private final int category;

        public InventoryAdapter(ArrayList<Pair<Item, Integer>> data) {
            this.data.addAll(data);
            category = 0;
        }

        public InventoryAdapter(ArrayList<Pair<Item, Integer>> data, int category) {
            this.category = category;
            this.data.addAll(data);
        }

        private class ViewHolder extends RecyclerView.ViewHolder {
            TextView name;

            public ViewHolder(@NonNull @NotNull View itemView) {
                super(itemView);
                name = (TextView) itemView.findViewById(R.id.textView);
            }
        }

        @NonNull
        @NotNull
        @Override
        public InventoryFragment.InventoryAdapter.ViewHolder onCreateViewHolder(@NonNull @NotNull ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.inventory_item, parent, false);
            return new ViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull @NotNull InventoryFragment.InventoryAdapter.ViewHolder holder, int position) {
            holder.name.setText(">" + data.get(position).first.getName() + ":" + data.get(position).second + "\n");
            holder.name.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (data.get(position).first.getCategory() == 0) {
                        MainActivity.player.equip(data.get(position).first);
                    }
                    FragmentManager fm = getChildFragmentManager();
                    FragmentTransaction fr = fm.beginTransaction();
                    if (fm.findFragmentById(R.id.characteristics) != null)
                        fr.remove(fm.findFragmentById(R.id.characteristics));
                    fr.add(R.id.characteristics, new ItemCharacteristicsFragment(data.get(position).first));
                    fr.commit();
                }
            });
        }

        @Override
        public int getItemCount() {
            if (category == -1){
                if (data.size() == 0)
                    noItems.setVisibility(View.VISIBLE);
                else noItems.setVisibility(View.GONE);
                return data.size();
            }
            else {
                ArrayList<Pair<Item, Integer>> data1 = new ArrayList<>();
                for (int i = 0; i < data.size(); i++) {
                    if (data.get(i).first.getCategory() == category)
                        data1.add(data.get(i));
                }
                data = new ArrayList<>(data1);
                if (data.size() == 0)
                    noItems.setVisibility(View.VISIBLE);
                else noItems.setVisibility(View.GONE);
                return data1.size();
            }
        }
    }
}

File: FinalProject/app/src/main/java/com/example/finalproject/fragments/InventoryFragment.kt
package com.example.finalproject.fragments

import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.finalproject.R
import com.example.finalproject.items.Item

class InventoryFragment : Fragment(), View.OnClickListener {

    private var noItems: TextView? = null
    private val categories = ArrayList<ImageView>()

    private lateinit var back: Button
    private lateinit var inventory: RecyclerView

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_inventory, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        noItems = requireView().findViewById(R.id.textView12)
        inventory = requireView().findViewById(R.id.list)
        back = requireView().findViewById(R.id.inventory_back_button)
        categories.add(requireView().findViewById(R.id.armor_weapons))
        categories.add(requireView().findViewById(R.id.potions_food))
        categories.add(requireView().findViewById(R.id.recourses))
        categories.add(requireView().findViewById(R.id.other))
        for (i in 0..3) {
            val bm: Bitmap = Bitmap.createScaledBitmap(
                MainActivity.textures[6][i],
                MainActivity.categoryImageWidth,
                MainActivity.categoryImageWidth,
                false
            )
            categories[i].setImageBitmap(Bitmap.createBitmap(bm))
            categories[i].setOnClickListener(this)
        }
        inventory.adapter = InventoryAdapter(MainActivity.player.inventory)
        inventory.layoutManager = LinearLayoutManager(context)
        back.setOnClickListener(this)
    }

    private inner class InventoryAdapter(
        val data: ArrayList<Pair<Item, Int>>,
        val category: Int = 0
    ) : RecyclerView.Adapter<InventoryAdapter.ViewHolder>() {

        private inner class ViewHolder(itemView: View) :
            RecyclerView.ViewHolder(itemView) {
            var name: TextView = itemView.findViewById(R.id.textView)
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val view =
                LayoutInflater.from(parent.context).inflate(R.layout.inventory_item, parent, false)
            return ViewHolder(view)
        }

        @SuppressLint("SetTextI18n")
        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            holder.name.text = ">${data[position].first.name}:${data[position].second}"
            holder.name.setOnClickListener {
                if (data[position].first.category == 0) {
                    MainActivity.player.equip(data[position].first)
                }
                val fm = childFragmentManager
                val fr = fm.beginTransaction()
                fm.findFragmentById(R.id.characteristics)?.let { it1 -> fr.remove(it1) }
                fr.add(R.id.characteristics, ItemCharacteristicsFragment(data[position].first))
                fr.commit()
            }
        }

        override fun getItemCount(): Int {
            return if (category == -1) {
                if (data.size == 0) noItems!!.visibility = View.VISIBLE else noItems!!.visibility =
                    View.GONE
                data.size
            } else {
                val data1 = ArrayList<Pair<Item, Int>>()
                for (i in data.indices) {
                    if (data[i].first.category == category) data1.add(data[i])
                }
                data.clear()
                data.addAll(data1)
                if (data.size == 0) noItems!!.visibility = View.VISIBLE else noItems!!.visibility =
                    View.GONE
                data1.size
            }
        }
    }

    override fun onClick(p0: View?) {
        if (p0 is ImageView) {
            inventory.adapter =
                InventoryAdapter(MainActivity.player.inventory, categories.indexOf(p0))
        } else if (p0 == back) {
            val fragmentManager = parentFragmentManager
            val fragmentTransaction = fragmentManager.beginTransaction()
            fragmentManager.findFragmentById(R.id.inventory)
                ?.let { it1 -> fragmentTransaction.remove(it1) }
            fragmentTransaction.add(R.id.map, MapFragment(MainActivity.player.mapNum))
            fragmentTransaction.add(R.id.status, StatusBarFragment())
            fragmentTransaction.add(R.id.menu, MenuFragment())
            fragmentTransaction.commit()
        }
    }
}